MULTIARCH-4282: add perfscale test job

qiliRedHat opened the pull request:
https://issues.redhat.com/browse/OCPQE-24888
https://issues.redhat.com/browse/MULTIARCH-4282

qiliRedHat commented:
/pj-rehearse periodic-ci-openshift-multiarch-tuning-operator-main-e2e-gcp-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
@aleskandro Could you help to check this job failure https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-periodic-ci-openshift-multiarch-tuning-operator-main-e2e-gcp-ovn-mto-heterogeneous-perfscale/1823974308272148480? 
multiarch-tuning-operator-deploy-commands.sh [failed](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-periodic-ci-openshift-multiarch-tuning-operator-main-e2e-gcp-ovn-mto-heterogeneous-perfscale/1823974308272148480/artifacts/e2e-gcp-ovn-mto-heterogeneous-perfscale/multiarch-tuning-operator-deploy/build-log.txt). Debug log said 'multiarch-tuning-operator-controller-manager' pods are in ContainerCreating status, while all 24 worker nodes are Ready. Would you like to add more debug info like the Event of the namespace and log of the pods?
```
Getting deployment info...

oc -n openshift-multiarch-tuning-operator get deployments -owide
NAME                                           READY   UP-TO-DATE   AVAILABLE   AGE   CONTAINERS                IMAGES                                                                                                                                                                               SELECTOR
multiarch-tuning-operator-controller-manager   0/2     0            0           3s    manager,kube-rbac-proxy   registry.ci.openshift.org/origin/multiarch-tuning-operator:main,gcr.io/kubebuilder/kube-rbac-proxy:v0.13.1@sha256:d4883d7c622683b3319b5e6b3a7edfbf2594c18060131a8bf64504805f875522   control-plane=controller-manager
Getting pod info....

oc -n openshift-multiarch-tuning-operator get pods -owide
NAME                                                           READY   STATUS              RESTARTS   AGE   IP       NODE                                                   NOMINATED NODE   READINESS GATES
multiarch-tuning-operator-controller-manager-c5cf4d4ff-6b4ld   0/2     ContainerCreating   0          0s    <none>   ci-op-ysgglmhc-b50ae-fmx8v-worker-a-additional-whg5d   <none>           <none>
multiarch-tuning-operator-controller-manager-c5cf4d4ff-fjd9p   0/2     ContainerCreating   0          0s    <none>   ci-op-ysgglmhc-b50ae-fmx8v-worker-f-wvc4v              <none>           <none>
Getting nodes info...
```

qiliRedHat commented:
/pj-rehearse periodic-ci-openshift-multiarch-tuning-operator-main-e2e-gcp-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-e2e-gcp-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
@aleskandro I added some more debug steps and re-run the job, but this time the Step e2e-gcp-ovn-mto-heterogeneous-XXXXXXXXX-multiarch-tuning-operator-deploy succeeded. It seems to be a temporary issue. Will see what debug steps get in the next time it occurs.

qiliRedHat commented:
@aleskandro I added some more debug steps and re-run the job, but this time the Step e2e-gcp-ovn-mto-heterogeneous-XXXXXXXXX-multiarch-tuning-operator-deploy succeeded. It seems to be a temporary issue. Will see what debug steps get in the next time it occurs.

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
Job https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-mto-heterogeneous-perfscale/1824269282482065408 e2e failed. Will skip ref: multiarch-tuning-operator-e2e

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-heterogeneous-perfscale

qiliRedHat commented:
The node density test result with MTO
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-mto-heterogeneous-perfscale/1824372462456934400
```
time="2024-08-16 10:30:58" level=info msg="node-density: Ready 50th: 107000 99th: 209000 max: 211000 avg: 106109" file="pod_latency.go:242"
time="2024-08-16 10:30:58" level=info msg="node-density: PodScheduled 50th: 106000 99th: 207000 max: 210000 avg: 104845" file="pod_latency.go:242"
time="2024-08-16 10:30:58" level=info msg="node-density: ContainersReady 50th: 107000 99th: 209000 max: 211000 avg: 106109" file="pod_latency.go:242"
time="2024-08-16 10:30:58" level=info msg="node-density: Initialized 50th: 106000 99th: 207000 max: 210000 avg: 104864" file="pod_latency.go:242"
time="2024-08-16 10:30:58" level=error msg="podLatency: P99 Ready latency (209.00s) higher than configured threshold: 3m0s" file="job.go:202"
```

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-heterogeneous-perfscale

qiliRedHat commented:
@aleskandro 
Please help to review this PR. 2 new jobs are added to run node-density tests, one job is with MTO deployed, the other is without. So we can compare the results.

First round of node-density tests are done. 
With MTO the PodScheduled time is much longer than without MTO. avg 104845 ms v.s. 224 ms, p99 207s v.s. 2s
The time from PodScheduled till Ready are similar.

> The node density test result with MTO: PodScheduled 50th: 106000 99th: 207000 max: 210000 avg: 104845
> 
> https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-mto-heterogeneous-perfscale/1824372462456934400
> 
> ```
> time="2024-08-16 10:30:58" level=info msg="node-density: Ready 50th: 107000 99th: 209000 max: 211000 avg: 106109" file="pod_latency.go:242"
> time="2024-08-16 10:30:58" level=info msg="node-density: PodScheduled 50th: 106000 99th: 207000 max: 210000 avg: 104845" file="pod_latency.go:242"
> time="2024-08-16 10:30:58" level=info msg="node-density: ContainersReady 50th: 107000 99th: 209000 max: 211000 avg: 106109" file="pod_latency.go:242"
> time="2024-08-16 10:30:58" level=info msg="node-density: Initialized 50th: 106000 99th: 207000 max: 210000 avg: 104864" file="pod_latency.go:242"
> time="2024-08-16 10:30:58" level=error msg="podLatency: P99 Ready latency (209.00s) higher than configured threshold: 3m0s" file="job.go:202"
> ```
> 
> The node density test result without MTO: PodScheduled 50th: 0 99th: 2000 max: 2000 avg: 224
> 
> https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-heterogeneous-perfscale/1826068194654162944
> 
> ```
> time="2024-08-21 02:24:06" level=info msg="node-density: PodScheduled 50th: 0 99th: 2000 max: 2000 avg: 224" file="pod_latency.go:242"
> time="2024-08-21 02:24:06" level=info msg="node-density: ContainersReady 50th: 1000 99th: 8000 max: 9000 avg: 2027" file="pod_latency.go:242"
> time="2024-08-21 02:24:06" level=info msg="node-density: Initialized 50th: 0 99th: 3000 max: 3000 avg: 331" file="pod_latency.go:242"
> time="2024-08-21 02:24:06" level=info msg="node-density: Ready 50th: 1000 99th: 8000 max: 9000 avg: 2027" file="pod_latency.go:242"
> ```

aleskandro commented:
/retitle WIP: MULTIARCH-4282: add perfscale test job

aleskandro commented on ci-operator/step-registry/multiarch/tuning-operator/deploy/multiarch-tuning-operator-deploy-commands.sh:
```diff
@@ -15,6 +15,17 @@ function debug() {
+            echo "Getting pod log of $name, status is $status"
+            oc logs -n $NAMESPACE $name
+          fi
+        done
```
Hi @qiliRedHat, With the recent changes, we are now exposing some labels that might help understand the operator's behavior at scale.

See https://github.com/openshift/multiarch-tuning-operator/pull/247

If I understand correctly, node density is meant to create a constant number of pods per second in the cluster under testing.

As part of the perf/scale efforts, we would like to know:


1. How many pods did get the scheduling gate set and removed? `multiarch.openshift.io/scheduling-gate: removed`
2. How many pods did NOT get the scheduling gate set? `multirach.openshift.io/scheduling-gate` shouldn't be set. This value should be 0.
3. How many pods did get the node affinity set? `multirach.openshift.io/node-affinity: set`. This value should be equal to the one in (1)
4. How many pods did NOT get the the node affinity set? This should be equal to (2)=0=(3)-(1)
5. Any left pending pod? `multiarch.openshift.io/scheduling-gate: gated` or the status.Phase == pending with reason "SchedulingGated".


If the pods are not deleted at the end of the testing, we might get these information here too.

aleskandro replied:
cc @lwan-wanglin

lwan-wanglin replied:
yes, we can add junits for that

qiliRedHat replied:
Yes, I think it is possible to configure to not let the pods being deleted at the end of the testing. I'll check how to do it and update the jobs.

qiliRedHat replied:
@aleskandro I have added a env `GC: "false"` to not deleting the test workload after the `openshift-qe-node-density` test.
@lwan-wanglin If you have added the junits to check the measures, please let me know what test I can add to the end of the perfscale test `openshift-qe-node-density`.

qiliRedHat replied:
@lwan-wanglin The label of the test pods is `kube-burner-job=node-density`. If possible we can make it a env var to be more portable to other tests.

lwan-wanglin replied:
talked with @qiliRedHat in private, I will take the following labels calculation and junits creation.

aleskandro commented on ci-operator/config/openshift/multiarch-tuning-operator/openshift-multiarch-tuning-operator-main.yaml:
```diff
@@ -273,6 +273,45 @@ tests:
+    - ref: ipi-install-heterogeneous
+    - ref: multiarch-tuning-operator-deploy
+    - ref: multiarch-tuning-operator-deploy-pod-placement
+    - ref: openshift-qe-node-density
```
I'm working on exposing the same information in https://github.com/openshift/release/pull/55603/files#r1744367854 as gauges/histogram Prometheus metrics and checking if we can onboard them in the cluster-monitoring operator. See https://issues.redhat.com/browse/MULTIARCH-4294


Based on our previous conversation, there should be a way to obtain the `time to schedule` (from the creation time to the pod having a value for .spec.nodeName). 

Is there also a way to catch custom metrics (coming from the Prometheus instance of the cluster-monitoring operator) in your grafana dashboards?

qiliRedHat replied:
I need some time to study how to gather 'customized' metrics and show them on grafana dashboard.

aleskandro commented on ci-operator/config/openshift/multiarch-tuning-operator/openshift-multiarch-tuning-operator-main.yaml:
```diff
@@ -273,6 +273,45 @@ tests:
+    - ref: multiarch-tuning-operator-deploy
+    - ref: multiarch-tuning-operator-deploy-pod-placement
+    - ref: openshift-qe-node-density
+    workflow: ipi-gcp-ovn
```
Is there a way to setup the number of pods/s we create (or similar)?

qiliRedHat replied:
Yes, it can be set by env [PODS_PER_NODE](https://github.com/openshift/release/blob/master/ci-operator/step-registry/openshift-qe/node-density/openshift-qe-node-density-ref.yaml#L32) , it is 245 by default if you don't set.

qiliRedHat replied:
Added env PODS_PER_NODE to the job as 245, it can be overwritten if needed.

qiliRedHat commented:
@aleskandro What do you think about the big difference of the PodScheduled time with and without MTO in my test above? Is there a 'expected' performance downgrade with MTO in your mind?

aleskandro commented:
> @aleskandro Please help to review this PR. 2 new jobs are added to run node-density tests, one job is with MTO deployed, the other is without. So we can compare the results.
> 
> First round of node-density tests are done. With MTO the PodScheduled time is much longer than without MTO. avg 104845 ms v.s. 224 ms, p99 207s v.s. 2s The time from PodScheduled till Ready are similar.
> 
> > The node density test result with MTO: PodScheduled 50th: 106000 99th: 207000 max: 210000 avg: 104845
> > https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-mto-heterogeneous-perfscale/1824372462456934400
> > ```
> > time="2024-08-16 10:30:58" level=info msg="node-density: Ready 50th: 107000 99th: 209000 max: 211000 avg: 106109" file="pod_latency.go:242"
> > time="2024-08-16 10:30:58" level=info msg="node-density: PodScheduled 50th: 106000 99th: 207000 max: 210000 avg: 104845" file="pod_latency.go:242"
> > time="2024-08-16 10:30:58" level=info msg="node-density: ContainersReady 50th: 107000 99th: 209000 max: 211000 avg: 106109" file="pod_latency.go:242"
> > time="2024-08-16 10:30:58" level=info msg="node-density: Initialized 50th: 106000 99th: 207000 max: 210000 avg: 104864" file="pod_latency.go:242"
> > time="2024-08-16 10:30:58" level=error msg="podLatency: P99 Ready latency (209.00s) higher than configured threshold: 3m0s" file="job.go:202"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The node density test result without MTO: PodScheduled 50th: 0 99th: 2000 max: 2000 avg: 224
> > https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-gcp-ovn-heterogeneous-perfscale/1826068194654162944
> > ```
> > time="2024-08-21 02:24:06" level=info msg="node-density: PodScheduled 50th: 0 99th: 2000 max: 2000 avg: 224" file="pod_latency.go:242"
> > time="2024-08-21 02:24:06" level=info msg="node-density: ContainersReady 50th: 1000 99th: 8000 max: 9000 avg: 2027" file="pod_latency.go:242"
> > time="2024-08-21 02:24:06" level=info msg="node-density: Initialized 50th: 0 99th: 3000 max: 3000 avg: 331" file="pod_latency.go:242"
> > time="2024-08-21 02:24:06" level=info msg="node-density: Ready 50th: 1000 99th: 8000 max: 9000 avg: 2027" file="pod_latency.go:242"
> > ```

The time-to-schedule should improve as soon as we commit https://issues.redhat.com/browse/MULTIARCH-4781.

In general, yes. The time to schedule is expected to increase, even though that 50th percentile seems high. The reason is that with the operator in place, we perform at least 1 HTTP request per image per pod in the pod-placement-controller. Also, the pod-placement-controller is stateful and doesn't scale. Therefore, we only have a single process consuming 'pods' for now.



Some thoughts:
1. Can we deploy the MTO in dedicated nodes to isolate it from the workloads so that we can double check what's going on with the median
2. Are the workloads using all the same image? What's the registry where the image is located? Can we use an image in the internal registry if it's outside the cluster? 
3. How many pods per second do we create when the test runs?
4.

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp418-gcp-ovn-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp418-aws-ovn-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-heterogeneous-perfscale

qiliRedHat commented:
Rebased the PR, moved the perfscale test jobs to 4.18.
[Test on 4.18](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-ocp418-aws-ovn-heterogeneous-perfscale/1843558732596776960) failed due to image registry issue. Asked the question here https://redhat-internal.slack.com/archives/CBN38N3MW/p1728379642621299
Moved the perfscale test jobs to 4.18. [Test on 4.17](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-heterogeneous-perfscale/1843898209269714944) also failed for the image registry issue.

deepsm007 commented:
https://github.com/openshift/release/blob/b83a6e10aa45eee1da0ddf4fdb80fee394fea23b/ci-operator/config/openshift/multiarch-tuning-operator/.config.prowgen

deepsm007 commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
/pj-rehearse pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-mto-heterogeneous-perfscale

qiliRedHat commented:
@aleskandro 2nd round of tests on AWS OVN with build "ocpVersion":"4.17.0"
The node density test result without MTO: PodScheduled 50th: 0 99th: 0 max: 1000 avg: 3
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-heterogeneous-perfscale/1844190056076546048
```
time="2024-10-10 02:53:27" level=info msg="node-density: Ready 50th: 1000 99th: 2000 max: 3000 avg: 1171" file="pod_latency.go:238"
time="2024-10-10 02:53:27" level=info msg="node-density: PodScheduled 50th: 0 99th: 0 max: 1000 avg: 3" file="pod_latency.go:238"
time="2024-10-10 02:53:27" level=info msg="node-density: ContainersReady 50th: 1000 99th: 2000 max: 3000 avg: 1171" file="pod_latency.go:238"
time="2024-10-10 02:53:27" level=info msg="node-density: Initialized 50th: 0 99th: 0 max: 1000 avg: 5" file="pod_latency.g
```
The node density test result with MTO: PodScheduled 50th: 0 99th: 1000 max: 1000 avg: 39
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_release/55603/rehearse-55603-pull-ci-openshift-multiarch-tuning-operator-main-ocp417-aws-ovn-mto-heterogeneous-perfscale/1844379713422757888
```
time="2024-10-10 16:00:25" level=info msg="node-density: ContainersReady 50th: 1000 99th: 2000 max: 3000 avg: 1297" file="pod_latency.go:238"
time="2024-10-10 16:00:25" level=info msg="node-density: Initialized 50th: 0 99th: 1000 max: 1000 avg: 52" file="pod_latency.go:238"
time="2024-10-10 16:00:25" level=info msg="node-density: Ready 50th: 1000 99th: 2000 max: 3000 avg: 1297" file="pod_latency.go:238"
time="2024-10-10 16:00:25" level=info msg="node-density: PodScheduled 50th: 0 99th: 1000 max: 1000 avg: 39" file="pod_latency.go:238"
```

qiliRedHat commented:
/test required

qiliRedHat commented:
/retest-required

qiliRedHat commented:
@aleskandro PTAL

aleskandro commented on ci-operator/config/openshift/multiarch-tuning-operator/openshift-multiarch-tuning-operator-main__ocp417.yaml:
```diff
@@ -123,7 +123,49 @@ tests:
     - ref: openshift-e2e-test
-    workflow: openshift-e2e-aws-ovn
+    workflow: ipi-aws-ovn
+- always_run: false
```
```suggestion
- optional: true
  run_if_changed: "\.perfscale$"
```

qiliRedHat replied:
@aleskandro This line failed the `make jobs`
```
time="2024-10-12T01:48:10Z" level=fatal msg="failed to parse arguments" error="failed to create CI Op config agent: failed to load config: loading config failed: failed to load ci-operator config (error converting YAML to JSON: yaml: line 128: found unknown escape character)"
```

qiliRedHat replied:
I used instead `run_if_changed: \.perfscale$` and the make jobs passed.
Or should I use `run_if_changed: "\\.perfscale$"`?

aleskandro replied:
Hi @qiliRedHat im not sure. Probably what you did is right... but we can change this later if it's not for sure

aleskandro commented on ci-operator/config/openshift/multiarch-tuning-operator/openshift-multiarch-tuning-operator-main__ocp417.yaml:
```diff
@@ -123,7 +123,49 @@ tests:
+    - ref: multiarch-tuning-operator-deploy-pod-placement
+    - ref: openshift-qe-node-density
+    workflow: ipi-aws-ovn
+- always_run: false
```
```suggestion
- optional: true
  run_if_changed: "\.perfscale$"
```

aleskandro reviewed (commented):
Hi @qiliRedHat, thanks for all this work!

Just a minor change to force the job to run only when we request it.

aleskandro commented:
/retitle "MULTIARCH-4282: add perfscale test job"

aleskandro commented:
/retitle MULTIARCH-4282: add perfscale test job

aleskandro commented:
/lgtm

qiliRedHat commented:
/pj-rehearse ack
